File: ios-native/BuilderHQ/Primitives/AuPhone.swift
/// AuPhone — Australian phone number validation, normalization, and
/// display formatting. Swift mirror of `src/lib/au-phone.ts` so the
/// client validates the same way the server does (no round-trip just
/// to learn a number is malformed).
///
/// Accepted input (whitespace / dashes / parens tolerated):
///   · Mobile:   04XX XXX XXX        → +614XXXXXXXX
///   · Landline: 0[2378] XXXX XXXX   → +61[2378]XXXXXXXX
///   · E.164:    +61… / 61…          → kept / +-prefixed
///
/// Valid leading significant digit (after the trunk 0 or +61): one of
/// 2, 3, 4, 7, 8.

import Foundation

enum AuPhone {
    private static let validLeading: Set<Character> = ["2", "3", "4", "7", "8"]

    /// Digits only, preserving a single leading `+`.
    private static func clean(_ input: String) -> String {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        let hasPlus = trimmed.hasPrefix("+")
        let digits = trimmed.filter(\.isNumber)
        return hasPlus ? "+\(digits)" : digits
    }

    /// Normalize to E.164 (`+61XXXXXXXXX`) or return nil if invalid.
    static func normalise(_ input: String) -> String? {
        let c = clean(input)

        // 0XXXXXXXXX — domestic trunk form.
        if c.count == 10, c.hasPrefix("0"), c.allSatisfy(\.isNumber) {
            let significant = String(c.dropFirst())
            guard let first = significant.first, validLeading.contains(first) else {
                return nil
            }
            return "+61\(significant)"
        }

        // +61XXXXXXXXX — already E.164.
        if c.hasPrefix("+61"), c.count == 12 {
            let significant = String(c.dropFirst(3))
            guard significant.allSatisfy(\.isNumber),
                  let first = significant.first,
                  validLeading.contains(first)
            else { return nil }
            return c
        }

        // 61XXXXXXXXX — E.164 without the plus.
        if c.hasPrefix("61"), c.count == 11, c.allSatisfy(\.isNumber) {
            let significant = String(c.dropFirst(2))
            guard let first = significant.first, validLeading.contains(first) else {
                return nil
            }
            return "+\(c)"
        }

        return nil
    }

    /// True when `input` is a valid AU mobile or landline.
    static func isValid(_ input: String) -> Bool {
        normalise(input) != nil
    }

    /// Convert a stored E.164 number (`+61XXXXXXXXX`) back to a
    /// friendly domestic display form (`04XX XXX XXX` / `0X XXXX
    /// XXXX`). Used to seed input fields on resume so the user sees
    /// the number the way they typed it, not the international form.
    /// Falls back to the raw input when it isn't a recognizable AU
    /// E.164 string.
    static func displayFromE164(_ stored: String) -> String {
        let trimmed = stored.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.hasPrefix("+61"), trimmed.count == 12 {
            let significant = trimmed.dropFirst(3)
            return formatAsTyping("0\(significant)")
        }
        return trimmed
    }

    /// Count of significant digits typed so far (trunk 0 / country code
    /// stripped). Used to decide when it's fair to show an error vs.
    /// "still typing".
    static func significantDigitCount(_ input: String) -> Int {
        let c = clean(input)
        if c.hasPrefix("+61") { return max(0, c.count - 3) }
        if c.hasPrefix("61"), c.count > 2 { return c.count - 2 }
        if c.hasPrefix("0") { return max(0, c.count - 1) }
        return c.filter(\.isNumber).count
    }

    /// Light display formatting as the user types. Groups a domestic
    /// number the way Australians read it:
    ///   · Mobile  04XX XXX XXX
    ///   · Landline 0X XXXX XXXX
    /// Leaves `+61…` and partial input largely untouched (just trims to
    /// digits + spaces) so we never fight the caret on edge cases.
    static func formatAsTyping(_ input: String) -> String {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        // Keep E.164 input as-is (digits + leading +) — formatting an
        // international form mid-type is more confusing than helpful.
        if trimmed.hasPrefix("+") {
            return "+" + trimmed.dropFirst().filter(\.isNumber)
        }

        let digits = String(trimmed.filter(\.isNumber).prefix(10))
        guard digits.hasPrefix("0"), digits.count > 1 else { return digits }

        let isMobile = digits.count >= 2 && digits[digits.index(digits.startIndex, offsetBy: 1)] == "4"
        let groups: [Int] = isMobile ? [4, 3, 3] : [2, 4, 4]

        var result = ""
        var idx = digits.startIndex
        for (gi, size) in groups.enumerated() {
            guard idx < digits.endIndex else { break }
            if gi > 0 { result += " " }
            let end = digits.index(idx, offsetBy: size, limitedBy: digits.endIndex) ?? digits.endIndex
            result += digits[idx..<end]
            idx = end
        }
        return result
    }
}

File: src/lib/au-phone.ts
/**
 * Australian phone number validation + normalization.
 *
 * Single source of truth for AU phone handling across the codebase.
 * Accepts the formats real people type and normalizes them to E.164
 * (`+61…`) for storage. The marketing-funnel routes had an inline
 * `normalisePhone` that only handled mobiles (04xx); this generalizes
 * it to mobiles AND landlines so onboarding can require a reachable
 * number of any kind.
 *
 * Accepted input shapes (whitespace / dashes / parens tolerated):
 *   · Mobile:   04XX XXX XXX            → +614XXXXXXXX
 *   · Landline: 0[2378] XXXX XXXX       → +61[2378]XXXXXXXX
 *   · E.164:    +61 4XX XXX XXX / +61 [2378]…  (kept, validated)
 *   · 61-prefixed without +:  614…/612… → +61…
 *
 * Valid AU geographic/mobile leading digits after the trunk `0`
 * (or after `+61`): 2 (NSW/ACT), 3 (VIC/TAS), 4 (mobile),
 * 7 (QLD), 8 (SA/WA/NT). All are 9 significant digits after the
 * leading 0 / country code.
 */

const VALID_LEADING = new Set(["2", "3", "4", "7", "8"]);

/** Strip everything except digits and a single leading `+`. */
function clean(input: string): string {
  const trimmed = input.trim();
  const hasPlus = trimmed.startsWith("+");
  const digits = trimmed.replace(/[^\d]/g, "");
  return hasPlus ? `+${digits}` : digits;
}

/**
 * Normalize an AU phone number to E.164 (`+61XXXXXXXXX`).
 * Returns `null` when the input isn't a valid AU number.
 */
export function normaliseAuPhone(input: string): string | null {
  const c = clean(input);

  // 0XXXXXXXXX — domestic trunk form (10 digits, leading 0).
  if (/^0\d{9}$/.test(c)) {
    const significant = c.slice(1); // drop trunk 0 → 9 digits
    if (!VALID_LEADING.has(significant[0]!)) return null;
    return `+61${significant}`;
  }

  // +61XXXXXXXXX — already E.164 (9 significant digits).
  if (/^\+61\d{9}$/.test(c)) {
    const significant = c.slice(3);
    if (!VALID_LEADING.has(significant[0]!)) return null;
    return c;
  }

  // 61XXXXXXXXX — E.164 without the plus.
  if (/^61\d{9}$/.test(c)) {
    const significant = c.slice(2);
    if (!VALID_LEADING.has(significant[0]!)) return null;
    return `+${c}`;
  }

  return null;
}

/** True when `input` is a valid AU mobile or landline number. */
export function isValidAuPhone(input: string): boolean {
  return normaliseAuPhone(input) !== null;
}

/**
 * Count of significant digits typed so far (trunk `0` / country code
 * stripped). Lets a form decide when it's fair to show an inline error
 * ("you've typed enough that this is clearly wrong") vs. staying quiet
 * while the user is mid-entry.
 */
export function significantAuDigitCount(input: string): number {
  const c = clean(input);
  if (c.startsWith("+61")) return Math.max(0, c.length - 3);
  if (c.startsWith("61") && c.length > 2) return c.length - 2;
  if (c.startsWith("0")) return Math.max(0, c.length - 1);
  return c.replace(/\D/g, "").length;
}

/** True specifically for AU mobile numbers (significant digits start 4). */
export function isAuMobile(input: string): boolean {
  const e164 = normaliseAuPhone(input);
  return e164 !== null && e164.startsWith("+614");
}

/**
 * Convert a stored E.164 number (`+61XXXXXXXXX`) back to friendly
 * domestic form (`0XXXXXXXXX`) for seeding an input on resume. Returns
 * the input unchanged when it isn't a recognizable AU E.164 string.
 */
export function displayAuPhoneFromE164(stored: string | null | undefined): string {
  if (!stored) return "";
  const trimmed = stored.trim();
  if (/^\+61\d{9}$/.test(trimmed)) {
    return `0${trimmed.slice(3)}`;
  }
  return trimmed;
}
